Make ScratchMemorySpace semi-regular

masterleinad opened the pull request:
**edited** Making `ScratchMemorySpace` easier to use and to reason about since could not think of any good reason not to do so.  The copy-assignment was intentionally deleted in the past but there was no comment explaining that choice.

---
~~These seem to not be defined.
Since there is another constructor deleting the private default constructor doesn't seem to be any effect. I could be convinced that the copy constructor should be deleted, though.~~

Don't use the pre-C++11 way to delete functions but use the rule of 5 instead and also allow move assignment/move construction.

masterleinad committed:
Remove unimplemented member functions in ScratchMemorySpace

dalg24 commented:
Of course they are not defined.  That's the old way to delete them.
What is the rational to revisit?

masterleinad commented:
I was trying to understand the Travis problems with #3306 and was stumbling over this class.

dalg24 commented:
Why did you close?  The issue with your PR is not the changes you propose per se, but how you present them.
You are correct that the default constructor is implicitly deleted because we define a constructor.
But you do not discuss the changes to the assignment operator and nothing neither in the title line nor in your description highlight that change.

dhollman commented on core/src/Kokkos_ScratchSpace.hpp:
```diff
@@ -78,9 +78,6 @@ class ScratchMemorySpace {
   mutable int m_offset;
   mutable int m_default_level;
 
-  ScratchMemorySpace();
-  ScratchMemorySpace& operator=(const ScratchMemorySpace&);
```
```  
  ScratchMemorySpace() = delete;
  ScratchMemorySpace& operator=(const ScratchMemorySpace&) = delete;
```

nliber reviewed (commented):
This is saying we want ScratchMemorySpace to be copy constructible but not copy assignable.  Is that correct?

I'd much prefer the Rule of 5 here, which also has the advantage of not needing to explicitly delete the default constructor.

masterleinad committed:
Define all special member functions for ScratchMemorySpace

masterleinad commented:
> This is saying we want ScratchMemorySpace to be copy constructible but not copy assignable. Is that correct?

Yes, we need it to be copy constructible for TeamMembers to be copy constructible.
I was trying to bring up the question if it makes sense to also make it copy assignable (since all the member functions are `const` anyway) but there is currently really no need.
Let's just keep the current behavior (and allow move assignment/move construction) following the rule of 5 as well.

nliber commented:
Given that all the member variables are either `int` or `char*`, if we have copy assignment, it would do the exact same thing as move assignment.  I'd rather see that and use the rule of 0.

masterleinad committed:
Use rule of zero for ScratchMemorySpace

masterleinad commented:
Since that's what I originally proposed, that's fine with me. :slightly_smiling_face:

nliber committed:
Initialize the member variables in ScratchMemorySpace so that
any calls on a default constructed ScratchMemorySpace get_shmem* don't invoke ub

masterleinad committed:
Merge pull request #2 from nliber/cleanup_scratch_memory_space

Initialize member variables for default constructed ScratchMemorySpace

Rombur commented:
You need to apply clang-format

masterleinad committed:
Add default constructor

masterleinad commented:
With the latest changes, `ScratchMemorySpace` is semiregular now.
